ouroboros/chainsync/types.go

```diff
@@ -487,9 +487,60 @@ func (tt TxOuts) FindByAssetID(assetID AssetID) (TxOut, bool) {
 	return TxOut{}, false
 }
 
+type Datums map[string]string
+
+func (d *Datums) UnmarshalJSON(i []byte) error {
+	if i == nil {
+		return nil
+	}
+
+	var raw map[string]interface{}
+	err := json.Unmarshal(i, &raw)
+	if err != nil {
+		return fmt.Errorf("unable to unmarshal as raw map: %w", err)
+	}
+
+	results := make(Datums, len(raw))
+	// for backwards compatibility, since ogmios switched Datum values from []byte to hex string
+	for k, v := range raw {
+		if hexString, ok := v.(string); ok {
+			results[k] = hexString
+		} else {
+			results[k] = hex.EncodeToString(v.([]byte))
```

Comment:
if we're unmarshalling from JSON, would this ever be a byte array?

Also, do we have any cases where this got stored as base64?

Reply:
No they were likely also base64 strings coming out of ogmios. But for whatever reason, we decided to parse them as `[]byte`: see [here](https://github.com/savaki/ogmigo/blob/main/ouroboros/chainsync/types.go#L480).

Yep, ddb stores binary data as base64 encoded strings, eg:
```
          "datums": {
            "M": {
              "1892e71421370568c6825ec9dd84480dd228c6587986d2b3b6e62305df4ea7b0": {
                "B": "2HmfQi0B2Hmf2Hmf2Hmf2HmfWBwBkEhF8GErPgi93ej3pmlmMM/0B1FRevcCTetu/9h5n9h5n9h5n1gcZBO+6DBVnEQ5DEy9KB3sESZUhzpRVKDKRAwitf/////YeoD/2HqA/xoAJiWg2Hmf2HmAGgfyToTYeZ8aL68IAP///w=="
              },
              "660769a83940cb0046a10329beba5a2a8f8ce172e9e24d4f4c7f46db46b6fd7d": {
                "B": "2HmfWBzfyDgAvlUG+aRN9FKqGM0/N5n//cXtyJ7cYYc2Qp4K/w=="
              },
              "7dd5dcc7f0223ce49a6b683c35ea96f5e5f6b8acf11275025f0d9fe7ce286336": {
                "B": "2Hmf2Hmf2HmfQED/2HmfWBxqyO8ztRDsAE/hFYX3xanwwH8MI0KKtPKcHX0QRE1FTET//0ItARsAAAFAVsJ+idh5nwMZA+j//w=="
              },
              "a221708d2ad3bfbc2a7e0672b694a315aadbefdd7d354af8ac0ee3f680ca0790": {
                "B": "2HmfQi0B2Hmf2Hmf2Hmf2HmfWBzvWHm9pXgxf2aMZw7eXgaRVs82DJrLaMxeY8lC/9h5n9h5n9h5n1gczGjaf5gcTUjXM8At1+ISO4ndYYmm1+OvkWrcyf/////YeoD/2HqA/xoAJiWg2Huf2Hqf2HmfGhe8FFcastBeAP////8="
              },
              "dff982af93b7b48c016a5ec242de8986a4be7850030e529234dac2979379de79": {
                "B": "2HmfQi0B2Hmf2Hmf2Hmf2HmfWBxA7HNCMZLeoFkHUJoanPT0CbrvBKAWodzgOouh/9h5n9h5n9h5n1gcO9a0aczkbI+HHrjvasf4spRiLIG76sRndpj8Cv/////YeoD/2HqA/xoAJiWg2Huf2Hqf2HmfGg7zP7gadZEY9/////8="
              },
              "ef77792f2e73cbfff689ef0babfe92a893761de046757c2d82422429072c6530": {
                "B": "2Hmf2Hmf2HmfQED/2HmfWBxqyO8ztRDsAE/hFYX3xanwwH8MI0KKtPKcHX0QRE1FTET//0ItARsAAAFAvSRH3th5nwMZA+j//w=="
              }
            }
          }...
          ```